### ISSUE TITLE
Delete row via request_generate

### ISSUE THREAD

**Issue by yogat3ch** [tidyverse/googlesheets4#30]
This is a continuation of the issue raised on the googlesheets package issue [#357](https://github.com/jennybc/googlesheets/issues/357) regarding deleting a row.

Setting the cells to "" causes gs_add_row to add any additional rows below the line with all "", so the solution is not ideal as those cells are considered filled by the cellfeed. Deleting a row seems like a straightforward function to have. 
As per the advice given in 357, I am attempting to produce this functionality using the request_generate function in googlesheets4.
I'm referencing the sheets_endpoints and the [Googlesheets API documentation for rows/columns](https://developers.google.com/sheets/api/samples/rowcolumn). 
Using the following [openly editable sheet](https://docs.google.com/spreadsheets/d/1hkneXU7o8It8VpWS1GFMqi7l8vhaxNM9rTf3ZWB6Xsk/edit?usp=sharing) I'm attempting to delete row 2

Here is a reprex:

```r
googlesheets4::request_generate(endpoint = "sheets.spreadsheets.values.batchUpdate", params = list(spreadsheetId = "1hkneXU7o8It8VpWS1GFMqi7l8vhaxNM9rTf3ZWB6Xsk", requests = list(deleteDimension = list(range = list(sheetId = 0, dimension = "ROWS", startIndex = 1, endIndex = 2)))))
```
I'm receiving an error as follows:

`Error: These parameters are unknown:
requests`

I'm unable to find much guidance on Stackoverflow or the issues thread here on how to properly use the request_generate function so I thought I'd open an issue here. I hope it's the appropriate forum to do so!

**Comment by jennybc** [comment]
There are lots more editing functions in this package now, in case you want to revisit this and get it working for you. There are many batch update requests. For example this one:

https://github.com/tidyverse/googlesheets4/blob/f5c17f73c4ad50d4f07716e56ef93156e651909d/R/sheets_write.R#L133-L141

You aren't packaging your request properly, i.e. the requests need to part of the params.

googlesheets4 will eventually gain more high-level functions, such as one to delete cells, but I had to start with other features first.

Hopefully you can write what you need in the meantime.

**Comment by yogat3ch** [comment]
Hi Jenny,
Thanks for the response to the question. I've been trying various iterations for about an hour but it seems like I'm still just spinning my wheels with this. 
I did catch that the endpoint was specified wrong in my example `sheets.spreadsheets.values.batchUpdate` vs what it should have been `sheets.spreadsheets.batchUpdate`. 

> You aren't packaging your request properly, i.e. the requests need to part of the params.

I think I did have the request packaged as part of the params, but possibly not? I was confused by this instruction - I rewrote the request with a nested syntax to make it a little clearer:
```
.req <- googlesheets4::request_generate(endpoint = "sheets.spreadsheets.batchUpdate",
                                        params = list(
                                          spreadsheetId = "1hkneXU7o8It8VpWS1GFMqi7l8vhaxNM9rTf3ZWB6Xsk",
                                          requests = list(deleteDimension = list(
                                            range = list(
                                              sheetId = 0,
                                              dimension = "ROWS",
                                              startIndex = 1,
                                              endIndex = 2)))))
```
I took a look at it in JSON for the request to see how it compared to the example in the [API Documentation](https://developers.google.com/sheets/api/samples/rowcolumn) and it looks nearly identical with the exception of the brackets around the integers:
```
requests = list(deleteDimension = list(
  range = list(
    sheetId = 0,
    dimension = "ROWS",
    startIndex = 1,
    endIndex = 2)))
jsonlite::toJSON(requests)
```
I made the request:
```
.val <- googlesheets4::request_make(.req)
httr::content(.val)
```
However, when I ran the code it returned a `400` with an error: _Invalid value at 'requests'_.
I tried passing the JSON as the request:
```
.req <- googlesheets4::request_generate(endpoint = "sheets.spreadsheets.batchUpdate",
                                        params = list(
                                          spreadsheetId = "1hkneXU7o8It8VpWS1GFMqi7l8vhaxNM9rTf3ZWB6Xsk",
                                          requests = jsonlite::toJSON(requests)))
.val <- googlesheets4::request_make(.req)
httr::content(.val)
```
and received the same error.

I went into the sheets_write.R linked above and saw that you had a subroutine in creating the request called `bureq_clear_sheet`.
I found that in the documentation, and noticed it develops the request using `new`:
```
list(repeatCell = new(
    "RepeatCellRequest",
    range = new("GridRange", sheetId = sheetId),
    fields = "*"
  ))
```
I assume this is the `base::new` that calls a class. I re-wrote the syntax of the request using this function:
```
.req <- googlesheets4::request_generate(endpoint = "sheets.spreadsheets.batchUpdate",
                                        params = list(
                                          spreadsheetId = "1hkneXU7o8It8VpWS1GFMqi7l8vhaxNM9rTf3ZWB6Xsk",
                                          requests = new(deleteDimension, list(
                                            range = new("GridRange",
                                                        sheetId = 0,
                                                        dimension = "ROWS",
                                                        startIndex = 1,
                                                        endIndex = 2)))))
```
However I get an error _“GridRange” is not a defined class_. 
I thought that maybe this class was defined somewhere in the package and I couldn't access it because I wasn't calling it from within the package scope. I searched the repo for `setMethod` , `setClass` and `setRefClass` but didn't find any instances.  

I think this one is over my head it seems 🤦 😞

**Comment by czeildi** [comment]
Hi @yogat3ch ,

I managed to delete rows with the following code:

```r
req <- googlesheets4::request_generate(
    endpoint = "sheets.spreadsheets.batchUpdate",
    params = list(
        spreadsheetId = YOUR_SPREADSHEET_ID,
        requests = list("deleteDimension" = list(
            "range" = list(
                sheetId = 0,
                dimension = "ROWS",
                startIndex = 4,
                endIndex = 5
            )
        ))
    )
)

googlesheets4::request_make(req)
```

It does not follow the internal conventions of `googlesheets4`, e.g. using `new` constructor but worked for me on my own spreadsheet. Also `gridRange` seems to define a rectangular range not whole rows or columns needed by the `deleteDimension` request (https://developers.google.com/sheets/api/samples/rowcolumn) 

I believe the `new` function above in your example is not the `base::new` but the new function defined within `googlesheets4`.

**Comment by jennybc** [comment]
Thanks @czeildi for providing a working example.

Yes, internally I use the internal `new()` because it does some (currently light) validation against the spec.

But for a one-off like this, it's fine to just construct the necessary list / JSON this way.

I will put a proper interface around this when I return to googlesheets4, but this is great to help people prototype stuff, if they need to delete rows programmatically.

**Comment by yogat3ch** [comment]
Hi @czeildi,
Thanks for figuring this out!
It's a bit puzzling because the only difference between that and the request I used on the 2nd attempt is the double quotes around the list names:
`"deleteDimension"` and `"range"`
It worked when I modified my request to have the double quotes as yours does. I then removed the double quotes and tried it again, and it actually worked as I had tried it in the 2nd attempt! I'm clueless as to why it didn't work 10 days ago when I first tried it, but perhaps something changed in the package dependencies? At least it's working now.

@jennybc Thanks for the clarification on `new()`

I've included this in the text of the pull request as well:

To make the task of deleting columns and rows simpler for other users of googlesheets4, I wrapped this into a function called delete_dimensions in PR #93 

I realize that the name delete_dimensions doesn't quite fit the current function naming taxonomy for `googlesheets4`, so I'm open to suggestions as to how this could be more appropriately named.

The function uses two internals declared above it in delete_dimensions.R. One of which is `letters_to_num2` which performs a similar function to `cellranger::letters_to_num` but accounts for character specifications of a series of columns ie `B:D` and will output two column index numbers in this case. I just wanted to point this one out due to it's similarity to `letters_to_num`.

**Of note:** I included tests in test-delete_dimensions.R which work when I test using a sheet I have access to. I modified the code in the pull request to be similar to some of the other existing tests and use `test_sheet_create` such that it *should* work when Travis CI runs those tests with the credentials for the `googlesheets4-testing` account. 

I did not include a modification in `NEWS.md` as per the contributing guidelines as I wasn't sure about the naming of the function yet. However, if you're ok with it as is, let me know and I can go ahead and do so.

I also used Roxygen `7.0.2.9000` to build the Rd files. However, running it with devtools::document changed all the manual files. However, I only committed those files pertaining to the functions in the PR. I also did not include the modification to the DESCRIPTION file that updates the RoxygenNote version at the bottom.
